### PR TITLE
RectPartition and Deformation fixes

### PR DESCRIPTION
**Pull request by adler-j**
Fixes issue #747, `RectPartition.__ne__` missing and also fixes what I was aiming to fix, more relaxed conditions for the template space and coordinate space.

### CODE REVIEW

File: odl/deform/linearized.py [hunk 1]
```diff
@@ -136,13 +136,20 @@ class LinDeformFixedTempl(Operator):
...
-    def __init__(self, template):
+    def __init__(self, template, domain=None):
         """Initialize a new instance.
 
         Parameters
         ----------
         template : `DiscreteLpElement`
             Fixed template that is to be deformed.
+        domain : power space of `DiscreteLp`
+            The space of all allowable coordinates in the deformation.
```
Comment (kohr-h) [inline odl/deform/linearized.py]: allowable?
allowed I'd say

File: odl/deform/linearized.py [hunk 2]
```diff
@@ -136,13 +136,20 @@ class LinDeformFixedTempl(Operator):
...
+            A `ProductSpace` of ``template.ndim`` copies of a function-space.
+            It must fulfill
+            ``domain[0].partition == template.space.partition``, so
+            this option is useful mainly when different interpolations
```
Comment (kohr-h) [inline odl/deform/linearized.py]: when using

File: odl/deform/linearized.py [hunk 3]
```diff
@@ -136,13 +136,20 @@ class LinDeformFixedTempl(Operator):
...
 
-    def __init__(self, template):
+    def __init__(self, template, domain=None):
         """Initialize a new instance.
 
         Parameters
         ----------
         template : `DiscreteLpElement`
             Fixed template that is to be deformed.
+        domain : power space of `DiscreteLp`
```
Comment (kohr-h) [inline odl/deform/linearized.py]: optional

File: odl/deform/linearized.py [hunk 4]
```diff
@@ -136,13 +136,20 @@ class LinDeformFixedTempl(Operator):
...
+            The space of all allowable coordinates in the deformation.
+            A `ProductSpace` of ``template.ndim`` copies of a function-space.
+            It must fulfill
+            ``domain[0].partition == template.space.partition``, so
+            this option is useful mainly when different interpolations
+            in the displacement and template.
```
Comment (kohr-h) [inline odl/deform/linearized.py]: Default?

File: odl/deform/linearized.py [hunk 5]
```diff
@@ -170,12 +177,34 @@ def __init__(self, template):
...
+            raise TypeError('`template.space` must be a `DiscreteLp`'
+                            'instance, got {!r}'.format(space))
         self.__template = template
-        super().__init__(domain=self.template.space.real_space.tangent_bundle,
+
+        if domain is None:
+            domain = self.template.space.real_space.tangent_bundle
+        else:
+            if not isinstance(domain, ProductSpace):
+                raise TypeError('`coord_space` must be a `ProductSpace` '
```
Comment (kohr-h) [inline odl/deform/linearized.py]: check error message

File: odl/deform/linearized.py [hunk 6]
```diff
@@ -170,12 +177,34 @@ def __init__(self, template):
...
+                                'instance, got {!r}'.format(domain))
+            if not domain.is_power_space:
+                raise TypeError('`coord_space` must be a power space, '
```
Comment (kohr-h) [inline odl/deform/linearized.py]: check

File: odl/deform/linearized.py [hunk 7]
```diff
@@ -170,12 +177,34 @@ def __init__(self, template):
...
+                                'got {!r}'.format(domain))
+            if not isinstance(domain[0], DiscreteLp):
+                raise TypeError('`coord_space[0]` must be a `DiscreteLp` '
```
Comment (kohr-h) [inline odl/deform/linearized.py]: check

File: odl/deform/linearized.py [hunk 8]
```diff
@@ -170,12 +177,34 @@ def __init__(self, template):
...
+                                'instance, got {!r}'.format(domain[0]))
+
+            if template.space.partition != domain[0].partition:
+                raise ValueError('`template.space.partition` not '
```
Comment (kohr-h) [inline odl/deform/linearized.py]: break like this
```python
    raise ValueError(
        '...')
```
That should require much less lines.

File: odl/deform/linearized.py (hunk 4 above)
Comment (kohr-h) [inline odl/deform/linearized.py]: You mention complex spaces in the other variant but not here.

File: odl/deform/linearized.py [hunk 9]
```diff
@@ -298,28 +334,30 @@ def __init__(self, displacement, templ_space=None):
...
             templ_space = displacement.space[0]
         else:
             if not isinstance(templ_space, DiscreteLp):
                 raise TypeError('`templ_space` must be a `DiscreteLp` '
                                 'instance, got {!r}'.format(templ_space))
-            if templ_space.real_space.tangent_bundle != displacement.space:
-                raise ValueError('`templ_space.real_space.tangent_bundle` not '
-                                 'equal to `displacement.space` ({} != {})'
+            if templ_space.partition != space[0].partition:
+                raise ValueError('`templ_space.partition` not '
```
Comment (kohr-h) [inline odl/deform/linearized.py]: see above

File: odl/deform/linearized.py [hunk 10]
```diff
@@ -268,8 +302,10 @@ def __init__(self, displacement, templ_space=None):
         templ_space : `DiscreteLp`, optional
             Template space on which this operator is applied, i.e. the
             operator domain and range. It must fulfill
-            ``domain.real_space.tangent_bundle == displacement.space``, so this
-            option is useful mainly for support of complex spaces.
+            ``templ_space[0].partition == displacement.space.partition``, so
+            this option is useful mainly for support of complex spaces and if
+            different interpolations should be used for the displacement and
```
Comment (kohr-h) [inline odl/deform/linearized.py]: for ~~the~~

File: odl/deform/linearized.py (hunk 4 above)
Comment (adler-j) [inline odl/deform/linearized.py]: It is not relevant here, giving a complex template is handled properly already without it.

**Comment by adler-j** [comment]
Updated by comments. merge?

**Comment by kohr-h** [comment]
Sure!

**Pull request merged**